File: src/test/java/no/nav/veilarbaktivitet/brukernotifikasjon/BrukernotifikasjonAsserts.java
package no.nav.veilarbaktivitet.brukernotifikasjon;

import lombok.SneakyThrows;
import no.nav.brukernotifikasjon.schemas.input.BeskjedInput;
import no.nav.brukernotifikasjon.schemas.input.DoneInput;
import no.nav.brukernotifikasjon.schemas.input.NokkelInput;
import no.nav.brukernotifikasjon.schemas.input.OppgaveInput;
import no.nav.doknotifikasjon.schemas.DoknotifikasjonStatus;
import no.nav.veilarbaktivitet.config.kafka.kafkatemplates.KafkaStringAvroTemplate;
import no.nav.veilarbaktivitet.mock_nav_modell.BrukerOptions;
import no.nav.veilarbaktivitet.mock_nav_modell.MockBruker;
import no.nav.veilarbaktivitet.mock_nav_modell.MockNavService;
import no.nav.veilarbaktivitet.person.Person;
import no.nav.veilarbaktivitet.util.KafkaTestService;
import org.apache.avro.specific.SpecificRecord;
import org.apache.commons.lang3.NotImplementedException;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.assertj.core.api.Assertions;
import org.springframework.kafka.support.SendResult;

import static no.nav.veilarbaktivitet.brukernotifikasjon.kvittering.EksternVarslingKvitteringConsumer.FEILET;
import static no.nav.veilarbaktivitet.brukernotifikasjon.kvittering.EksternVarslingKvitteringConsumer.FERDIGSTILT;
import static no.nav.veilarbaktivitet.util.KafkaTestService.DEFAULT_WAIT_TIMEOUT_DURATION;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.kafka.test.utils.KafkaTestUtils.getSingleRecord;

public class BrukernotifikasjonAsserts {
    Consumer<String, String> brukernotifikasjonConsumer;
    private final KafkaStringAvroTemplate<DoknotifikasjonStatus> kviteringsProducer;

    BrukernotifikasjonAssertsConfig config;
    KafkaTestService kafkaTestService;

    public static MockBruker getBrukerSomIkkeKanVarsles() {
        return MockNavService.createBruker(BrukerOptions.happyBrukerBuilder().harBruktNivaa4(false).build());
    }

    public BrukernotifikasjonAsserts(BrukernotifikasjonAssertsConfig config) {
        kviteringsProducer = config.getKviteringsProducer();
        kafkaTestService = config.getTestService();
        brukernotifikasjonConsumer = config.createBrukernotifikasjonConsumer();
        this.config = config;
    }

    public void assertOppgaveSendt(Person.Fnr fnr) {
        getOppgaveSendt(fnr);
    }

    public String getOppgaveSedtId(Person.Fnr fnr) {
        return getOppgaveSendt(fnr).key();
    }

    public ConsumerRecord<String, String> getOppgaveSendt(Person.Fnr fnr) {
        config.getSendBrukernotifikasjonCron().sendBrukernotifikasjoner();

        ConsumerRecord<String, String> singleRecord = getSingleRecord(brukernotifikasjonConsumer, config.getBrukernotifikasjonTopic(), DEFAULT_WAIT_TIMEOUT_DURATION);


        String key = singleRecord.key();
        // TODO get fnr from body assertEquals(fnr.get(), key.getFodselsnummer());

        return singleRecord;
    }

    public ConsumerRecord<NokkelInput, BeskjedInput> assertBeskjedSendt(Person.Fnr fnr) {
        config.getSendBrukernotifikasjonCron().sendBrukernotifikasjoner();
        ConsumerRecord<NokkelInput, BeskjedInput> singleRecord = getSingleRecord(beskjedConsumer, config.getBeskjedTopic(), DEFAULT_WAIT_TIMEOUT_DURATION);

        NokkelInput key = singleRecord.key();
        assertEquals(fnr.get(), key.getFodselsnummer());
        assertEquals(config.getAppname(), key.getAppnavn());
        assertEquals(config.getNamespace(), key.getNamespace());

        return singleRecord;
    }

    public void assertIngenNyeBeskjeder() {
        config.getSendBrukernotifikasjonCron().sendBrukernotifikasjoner();
        Assertions.assertThat(kafkaTestService.harKonsumertAlleMeldinger(config.getBeskjedTopic(), beskjedConsumer))
                .as("Skal ikke sendes melding")
                .isTrue();
    }


    public ConsumerRecord<NokkelInput, DoneInput> assertDone(NokkelInput eventNokkel) {
        //Trigger scheduld jobb manuelt da schedule er disabled i test.
        config.getAvsluttBrukernotifikasjonCron().avsluttBrukernotifikasjoner();
        ConsumerRecord<NokkelInput, DoneInput> singleRecord = getSingleRecord(doneInputConsumer, config.getBrukernotifkasjonFerdigTopic(), DEFAULT_WAIT_TIMEOUT_DURATION);
        NokkelInput key = singleRecord.key();
        assertEquals(eventNokkel.getFodselsnummer(), key.getFodselsnummer());
        assertEquals(eventNokkel.getEventId(), key.getEventId());
        assertEquals(config.getAppname(), key.getAppnavn());
        assertEquals(config.getNamespace(), key.getNamespace());

        return singleRecord;
    }


    public void sendEksternVarseltOk(ConsumerRecord record) {
        sendVarsel(record, FERDIGSTILT);
    }

    public void sendEksternVarseletFeilet(ConsumerRecord record) {
        sendVarsel(record, FEILET);
    }

    @SneakyThrows
    public void sendVarsel(ConsumerRecord<NokkelInput, SpecificRecord> record, String status) {
        String kviteringsId = getKviteringsId(record);
        DoknotifikasjonStatus doknot = doknotifikasjonStatus(kviteringsId, status);
        SendResult<String, DoknotifikasjonStatus> result = kviteringsProducer.send(config.getKviteringsToppic(), kviteringsId, doknot).get();
        kafkaTestService.assertErKonsumert(config.getKviteringsToppic(), result.getRecordMetadata().offset());
    }

    public void assertSkalIkkeHaProdusertFlereMeldinger() {
        config.getAvsluttBrukernotifikasjonCron().avsluttBrukernotifikasjoner();
        config.getSendBrukernotifikasjonCron().sendBrukernotifikasjoner();

        kafkaTestService.harKonsumertAlleMeldinger(config.getOppgaveTopic(), oppgaveConsumer);
        kafkaTestService.harKonsumertAlleMeldinger(config.getBrukernotifkasjonFerdigTopic(), doneInputConsumer);
        kafkaTestService.harKonsumertAlleMeldinger(config.getBeskjedTopic(), beskjedConsumer);
    }


    private DoknotifikasjonStatus doknotifikasjonStatus(String bestillingsId, String status) {
        return DoknotifikasjonStatus
                .newBuilder()
                .setStatus(status)
                .setBestillingsId(bestillingsId)
                .setBestillerId("veilarbaktivitet")
                .setMelding("her er en melding")
                .setDistribusjonId(1L)
                .build();
    }

    private String getKviteringsId(ConsumerRecord<NokkelInput, SpecificRecord> record) {
        var valueClass = record.value().getClass();
        if (BeskjedInput.class.equals(valueClass)) {
            return "B-" + config.getAppname() + "-" + record.key().getEventId();
        }
        if(OppgaveInput.class.equals(valueClass)) {
            return "O-" + config.getAppname() + "-" + record.key().getEventId();
        }
        if(DoneInput.class.equals(valueClass)) {
            throw new IllegalArgumentException("Ikke nokk informasjon i denne til og finne kviterings id");
        }
        if(DoneInput.class.getPackage().equals(valueClass.getPackage())) {
            throw new NotImplementedException("ukjent klasse fra brukernotifikasjoner");
        }
        throw new IllegalArgumentException("Kommer denne klassen fra brukernotifikasjoner?");
    }
}

File: src/test/kotlin/no/nav/veilarbaktivitet/brukernotifikasjon/BrukernotifikasjonAsserts.kt
package no.nav.veilarbaktivitet.brukernotifikasjon

import junit.framework.TestCase.assertEquals
import no.nav.tms.varsel.action.Varseltype
import no.nav.veilarbaktivitet.brukernotifikasjon.opprettVarsel.InaktiverVarselDto
import no.nav.veilarbaktivitet.brukernotifikasjon.opprettVarsel.MinSideVarselId
import no.nav.veilarbaktivitet.brukernotifikasjon.varselStatusHendelse.EksternStatusOppdatertEventName
import no.nav.veilarbaktivitet.brukernotifikasjon.varselStatusHendelse.EksternVarselKanal
import no.nav.veilarbaktivitet.brukernotifikasjon.varselStatusHendelse.EksternVarselStatus
import no.nav.veilarbaktivitet.brukernotifikasjon.varselStatusHendelse.InternVarselHendelseType
import no.nav.veilarbaktivitet.person.Person.Fnr
import no.nav.veilarbaktivitet.util.KafkaTestService
import org.assertj.core.api.Assertions.assertThat
import java.util.*

class BrukernotifikasjonAsserts(var config: BrukernotifikasjonAssertsConfig) {
    var brukervarselConsumer = config.createBrukerVarselConsumer(config.testService)
    private val brukernotifikasjonVarselHendelseProducer = config.brukernotifikasjonVarselProducer()

    var kafkaTestService: KafkaTestService = config.testService

    private fun assertVarselSendt(fnr: Fnr, varselType: Varseltype): OpprettVarselDto {
        config.sendMinsideVarselFraOutboxCron.sendBrukernotifikasjoner()
        val varsel = brukervarselConsumer.waitForOpprettVarsel()
        assertEquals(fnr.get(), varsel.ident)
        assertEquals(varselType, varsel.type)
        assertEquals(config.appname, varsel.produsent.getAppnavn())
        assertEquals(config.namespace, varsel.produsent.getNamespace())
        return varsel
    }
    fun assertOppgaveSendt(fnr: Fnr) = assertVarselSendt(fnr, Varseltype.Oppgave)
    fun assertBeskjedSendt(fnr: Fnr) = assertVarselSendt(fnr, Varseltype.Beskjed)

    fun assertIngenNyeVarslerErIOutbox() {
        assertThat(config.sendMinsideVarselFraOutboxCron.sendAlle(500))
            .isEqualTo(0);
    }

    fun assertInaktivertMeldingErSendt(varselId: String?): InaktiverVarselDto {
        //Trigger scheduld jobb manuelt da schedule er disabled i test.
        config.avsluttBrukernotifikasjonCron.avsluttBrukernotifikasjoner()
        val inaktivering = brukervarselConsumer.waitForInaktiverVarsel()
        assertEquals(varselId, inaktivering.varselId)
        assertEquals(config.appname, inaktivering.produsent.appnavn)
        assertEquals(config.namespace, inaktivering.produsent.namespace)
        return inaktivering
    }

    fun simulerEksternVarselStatusSendt(varsel: OpprettVarselDto) {
        simulerEksternVarselStatusHendelse(
            MinSideVarselId(UUID.fromString(varsel.varselId)),
            EksternVarselStatus.sendt,
            varsel.type,
            kanal = EksternVarselKanal.SMS
        )
    }

    fun simulerEksternVarselStatusFeilet(varsel: OpprettVarselDto) {
        simulerEksternVarselStatusHendelse(
            MinSideVarselId(UUID.fromString(varsel.varselId)),
            EksternVarselStatus.feilet,
            varsel.type,
            null
        )
    }

    fun simulerEksternVarselStatusKansellert(varsel: OpprettVarselDto) {
        simulerEksternVarselStatusHendelse(
            MinSideVarselId(UUID.fromString(varsel.varselId)),
            EksternVarselStatus.kansellert,
            varsel.type,
            null
        )
    }

    fun simulerEksternVarselStatusFerdigstilt(varsel: OpprettVarselDto) {
        simulerEksternVarselStatusHendelse(
            MinSideVarselId(UUID.fromString(varsel.varselId)),
            EksternVarselStatus.ferdigstilt,
            varsel.type,
            null
        )
    }

    private fun simulerEksternVarselStatusHendelse(varselId: MinSideVarselId, status: EksternVarselStatus, varselType: Varseltype, kanal: EksternVarselKanal?) {
        val eksternVarsel = eksternVarselHendelse(varselId, status, varselType, kanal)
        simulerVarselHendelse(varselId, eksternVarsel)
    }

    fun simulerInternVarselStatusHendelse(varselId: MinSideVarselId, internVarselHendelseType: InternVarselHendelseType, varselType: Varseltype) {
        val internVarsel = internVarselHendelse(varselId, internVarselHendelseType, varselType)
        simulerVarselHendelse(varselId, internVarsel)
    }

    fun simulerVarselFraAnnenApp() {
        val annenAppSittVarsel = internVarselHendelse(MinSideVarselId(UUID.randomUUID()), InternVarselHendelseType.opprettet, Varseltype.Beskjed)
            .copy(appnavn = "annen app")
        simulerVarselHendelse(MinSideVarselId(annenAppSittVarsel.varselId), annenAppSittVarsel)
    }

    private fun simulerVarselHendelse(varselId: MinSideVarselId, hendelse: TestVarselHendelseDTO) {
        val result = brukernotifikasjonVarselHendelseProducer.publiserBrukernotifikasjonVarselHendelse(
            varselId, hendelse
        )
        kafkaTestService.assertErKonsumert(brukernotifikasjonVarselHendelseProducer.topic, result.recordMetadata.offset())
    }

    fun assertSkalIkkeHaProdusertFlereMeldinger() {
        config.avsluttBrukernotifikasjonCron.avsluttBrukernotifikasjoner()
        config.sendMinsideVarselFraOutboxCron.sendBrukernotifikasjoner()
        kafkaTestService.harKonsumertAlleMeldinger(config.brukernotifikasjonBrukervarselTopic, brukervarselConsumer.consumer)
    }

    private fun eksternVarselHendelse(varselId: MinSideVarselId, status: EksternVarselStatus, varselType: Varseltype, kanal: EksternVarselKanal?): EksternVarselHendelseDTO {
        return EksternVarselHendelseDTO(
            namespace = config.namespace,
            appnavn = config.appname,
            eventName = EksternStatusOppdatertEventName,
            status = status,
            varselId = varselId.value,
            varseltype = varselType,
            kanal = kanal
        )
    }

    private fun internVarselHendelse(varselId: MinSideVarselId, hendelseType: InternVarselHendelseType, varselType: Varseltype): InternVarselHendelseDTO {
        return InternVarselHendelseDTO(
            namespace = config.namespace,
            appnavn = config.appname,
            varseltype = varselType,
            varselId = varselId.value,
            eventName = hendelseType.name,
        )
    }
}
